Replaced existing threshold function with Otsu threshold algorithm

claudiofelber opened the pull request:
Hi Daniel

I replaced the existing threshold function with another one based on [Otsu's algorithm](https://en.wikipedia.org/wiki/Otsu%27s_method). Not only does this improve the overall quality of QR code detection, the whole detection process is also considerably faster. On my Intel desktop system detection time improved by a factor of 2.5 on average (even more for images with a lot of noise or bad lighting).

Would you mind having a look at my pull request?

Regards,
Claudio

**Example**
*Original image:*
![grayscale](https://user-images.githubusercontent.com/339950/56959298-3c4f6200-6b4d-11e9-8cde-089416ce10e9.jpg)

*Binarization with old threshold function:*
![binarized-old](https://user-images.githubusercontent.com/339950/56959332-64d75c00-6b4d-11e9-9372-dfe7abea3aa0.png)

*Binarization with new Otsu-based threshold function:*
![binarized-new](https://user-images.githubusercontent.com/339950/56959342-6dc82d80-6b4d-11e9-8f70-689b58e4bfe2.png)

kAworu commented:
Hi @claudiofelber :)

The patch looks good to me and your example is neat. I just have a couple of question on the code, see inline comments.

kAworu commented on lib/identify.c:
```diff
@@ -174,60 +174,55 @@ static void flood_fill_seed(struct quirc *q, int x, int y, int from, int to,
-#define THRESHOLD_T		5
-
-static void threshold(struct quirc *q)
+uint8_t otsu(struct quirc *q)
```
```suggestion
uint8_t otsu(const struct quirc *q)
```

kAworu commented on lib/identify.c:
```diff
@@ -174,60 +174,55 @@ static void flood_fill_seed(struct quirc *q, int x, int y, int from, int to,
+
+    // Calculate histogram
+    const int HISTOGRAM_SIZE = 256;
+    unsigned int histogram[HISTOGRAM_SIZE];
```
Now that a histogram is used, I guess the `row_average` member can be removed from `struct quirc` correct?

claudiofelber commented:
Hi Alexandre

Thank you for your feedback. You are absolutely right, the `row_average` member of `struct quirc` is no longer used. I have removed it including all associated code. I have also made the `struct quirc` parameter of the otsu function constant and fixed the indentation to match the one used in the remaining code (tabs instead of spaces).

dlbeer commented:
Looks good to me -- thanks for this!
